x-pack/plugins/observability/public/pages/slo_edit/components/custom_kql/index_selection.tsx

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useEffect, useState } from 'react';
+import { Control, Controller } from 'react-hook-form';
+import { EuiComboBox, EuiComboBoxOptionOption, EuiFormRow } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { CreateSLOInput } from '@kbn/slo-schema';
+
+import { useFetchIndices, Index } from '../../../../hooks/use_fetch_indices';
+
+export interface Props {
+  control: Control<CreateSLOInput>;
+}
+
+interface Option {
+  label: string;
+  options: Array<{ value: string; label: string }>;
+}
+
+export function IndexSelection({ control }: Props) {
+  const { loading, indices = [] } = useFetchIndices();
+  const [indexOptions, setIndexOptions] = useState<Option[]>([]);
+
+  useEffect(() => {
+    setIndexOptions([createIndexOptions(indices)]);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [indices.length]);
+
+  const onSearchChange = (search: string) => {
+    const options: Option[] = [];
+    if (!search) {
+      return setIndexOptions([createIndexOptions(indices)]);
+    }
+
+    const searchPattern = search.endsWith('*') ? search.substring(0, search.length - 1) : search;
+    const matchingIndices = indices.filter(({ name }) => name.includes(searchPattern));
+
+    if (matchingIndices.length === 0) {
+      return setIndexOptions([]);
+    }
+
+    options.push(createIndexOptions(matchingIndices));
+
+    const searchWithStarSuffix = search.endsWith('*') ? search : `${search}*`;
+    options.push({
+      label: i18n.translate(
+        'xpack.observability.slos.sloEdit.customKql.indexSelection.indexPatternLabel',
+        { defaultMessage: 'Use an index pattern' }
+      ),
+      options: [{ value: searchWithStarSuffix, label: searchWithStarSuffix }],
+    });
+
+    setIndexOptions(options);
+  };
+
+  return (
+    <EuiFormRow
+      label={i18n.translate('xpack.observability.slos.sloEdit.customKql.indexSelection.label', {
+        defaultMessage: 'Index',
+      })}
+      helpText={i18n.translate(
+        'xpack.observability.slos.sloEdit.customKql.indexSelection.helpText',
+        {
+          defaultMessage: 'Use * to broaden your query.',
+        }
+      )}
+    >
+      <Controller
+        name="indicator.params.index"
+        control={control}
+        rules={{
+          required: true,
+          validate: (value) => !!value,
```

Comment:
If you're just checking if any value is present, then `required: true` would suffice.

I do wonder if a more stringent validation function would be good here. One that checks if the input is at least partially part of any of the values returned by the useIndices hook. Theoretically, wouldn't it be possible to type in something in the box that does not correspond to any index pattern, and the form would treat it as a valid entry?

Reply:
Does `required:true` allow empty values, like `""`?

I think it does not hurt to add more validation to the index pattern indeed. I'll make some change

Reply:
[React Hook Form's docs](https://react-hook-form.com/api/useform/register) say this:

"[About `required`:] A Boolean which, if true, indicates that the input must have a value before the form can be submitted. You can assign a string to return an error message in the errors object." 

So setting it to true should not allow passing an empty string.

Reply:
addressed in https://github.com/elastic/kibana/pull/149786/commits/a9c43c3837b14eb545015bee9ed189954ad50950